mbq/atomiq/producers.py

```diff
@@ -118,6 +119,9 @@ def _create_task(self, task, *args, **kwargs):
                 'Atomiq Celery publish expects a "task" argument with a string "name" attribute.'
             )
 
+        if settings.ATOMIQ.get('synchronous', False):
+            task(*args, **kwargs)
+
         return models.CeleryTask.objects.create(
```

Comment:
I decided to keep the task creation here because (a) the BaseProducer uses the UUID of the created record, and (b) it's useful to exercise the JSONification of the payload in test.

Reply:
In that case I wonder if the task exection should happen after task creation, since that's more similar to how it works in prod, and will surface serialization errors before errors in the task code.

Reply:
that's a good point. i can do that.

Reply:
I hate to say it but it seems like the right implementation is to create the task and then run it in an on_commit 😱

Reply:
I agree with that idea, but I'm not sure if it works. I'm writing up a little test and I'll get back to you.

Reply:
here's what i've got:

```python3
# test.py
import unittest
from typing import Callable
from unittest.mock import Mock

from django.conf import settings
from django.db import transaction
from django.test import TestCase


settings.configure(DATABASES={
    'default': {
        'ENGINE': 'django.db.backends.sqlite3',
        'NAME': 'testdb',
    },
})


def do_it(fn: Callable[[], None], inline: bool) -> None:
    with transaction.atomic():
        if inline:
            fn()
        else:
            transaction.on_commit(fn)


class TestDoWithinTransaction(TestCase):
    def test_inline(self):
        fn = Mock()
        do_it(fn, inline=True)
        fn.assert_called_once_with()

    def test_on_commit(self):
        fn = Mock()
        do_it(fn, inline=False)
        fn.assert_called_once_with()


if __name__ == '__main__':
    unittest.main()
```

```
% python3 test.py
.F
======================================================================
FAIL: test_on_commit (__main__.TestDoWithinTransaction)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "test.py", line 35, in test_on_commit
    fn.assert_called_once_with()
  File "/usr/local/Cellar/python/3.7.0/Frameworks/Python.framework/Versions/3.7/lib/python3.7/unittest/mo
ck.py", line 830, in assert_called_once_with
    raise AssertionError(msg)
AssertionError: Expected 'mock' to be called once. Called 0 times.

----------------------------------------------------------------------
Ran 2 tests in 0.010s

FAILED (failures=1)
```

Reply:
I think this because execution of `on_commit` hooks is delayed until the outermost transaction has exited. Inside each `django.test.TestCase.test_*`, we still have the outer transaction set up for us to clean up test data.

Reply:
Oh true, yeah the testcase transaction will screw it up. Agree that on_commit probs won't work then 👍